src/compute/src/sink/copy_to_s3_oneshot.rs

```diff
@@ -67,68 +81,266 @@ where
 }
 
+// Required state to upload batches to S3
+struct CopyToS3Uploader {
+    // The output description.
+    desc: RelationDesc,
+    // Params to format the data.
+    format: CopyFormatParams<'static>,
+    // The index of the next file.
+    file_index: usize,
+    // The prefix for the file names.
+    file_name_prefix: String,
+    // The bucket prefix batches should be uploaded to.
+    bucket: String,
+    //The path prefix.
+    path_prefix: String,
+    // The desired batch size. When the buffer exceeds this value a put operation will be
+    // performed.
+    max_file_size: u64,
+    // The aws sdk config.
+    sdk_config: SdkConfig,
+    // Keeping the uploader in an `Option` to later take owned value.
+    single_file_uploader: Option<S3MultiPartUploader>,
+    // temporary buffer
+    buf: Vec<u8>,
+}
+
+impl CopyToS3Uploader {
+    fn new(
+        sdk_config: SdkConfig,
+        sink_connection: CopyToS3OneshotSinkConnection,
+        file_name_prefix: String,
+    ) -> CopyToS3Uploader {
+        let (bucket, path_prefix) = Self::extract_s3_bucket_path(&sink_connection.prefix);
+        CopyToS3Uploader {
+            desc: sink_connection.desc,
+            sdk_config,
+            format: sink_connection.format,
+            file_name_prefix,
+            bucket,
+            path_prefix,
+            max_file_size: sink_connection.max_file_size,
+            file_index: 0,
+            single_file_uploader: None,
+            buf: Vec::new(),
+        }
+    }
+
+    async fn start_new_file_upload(&mut self) -> Result<(), anyhow::Error> {
+        self.file_index += 1;
+        let file_path = self.file_path();
+        let bucket = self.bucket.clone();
+        info!(
+            "starting upload at bucket: {}, file {}",
+            &bucket, &file_path
+        );
+        let uploader = S3MultiPartUploader::try_new(
+            &self.sdk_config,
+            bucket,
+            file_path,
+            S3MultiPartUploaderConfig {
+                part_size_limit: ByteSize::mib(10).as_u64(),
+                file_size_limit: self.max_file_size,
+            },
+        )
+        .await?;
+        self.single_file_uploader = Some(uploader);
+        Ok(())
+    }
+
+    fn extract_s3_bucket_path(prefix: &str) -> (String, String) {
+        let uri = Uri::from_str(prefix).expect("valid s3 url");
+        let bucket = uri.host().expect("s3 bucket");
+        let path = uri.path().trim_start_matches('/').trim_end_matches('/');
+        (bucket.to_string(), path.to_string())
+    }
+
+    fn file_path(&self) -> String {
+        format!(
+            "{}/{}-{:04}.csv",
+            self.path_prefix, self.file_name_prefix, self.file_index
+        )
+    }
+
+    /// Finishes any remaining in progress upload.
+    async fn finish_ongoing_upload(&mut self) -> Result<(), anyhow::Error> {
+        if let Some(uploader) = self.single_file_uploader.take() {
+            uploader.finish().await?;
+        }
+        Ok(())
+    }
```

Comment:
Recommend rolling this into `start_new_ongoing_upload`. One less method to call!

Reply:
This method is also called when we reach the end in timely. Used to flush any ongoing upload before returning the count to the user.

Reply:
Ah, I see!